pkg/controller/clusterdeployment/clusterdeployment_controller.go

```diff
@@ -1794,10 +1801,16 @@ func generateDeprovision(cd *hivev1.ClusterDeployment) (*hivev1.ClusterDeprovisi
 
 	switch {
 	case cd.Spec.Platform.AWS != nil:
+		// If we haven't discovered the HostedZoneRole yet, fail
+		hzrp := controllerutils.AWSHostedZoneRole(cd)
```

Comment:
This role wont exit for all clusters, only those installing with the shared VPC. If I understand this correctly this will block all clusters deprovisions where the hosted zone role doesn't exist?

Reply:
This will only block deprovisions where the hosted zone role is unset **in the `cd.spec.clusterMetadata`**. We set that value in one of two code paths:
- When provision completes, we pull it out of the metadata.json laid down by the installer.
- For retrofitting existing clusters, we pull it out of the install-config.

In both of those installer structures, the field is of type `string`, not `*string` so it will always "exist".

However, `cd.spec.clusterMetadata.aws.hostedZoneRole` is of type `*string`, and we're distinguishing among three possible permutations:
- `nil` means "unset" -- i.e. we have never set the clusterMetadata field. This is the condition where we block deprovision.
- non-`nil`, empty string: this will be the case for legacy/non-shared-VPC clusters. This is a valid value and will *pass* the check on this line, allowing deprovision.
- non-`nil`, nonempty string: this is the case for shared-VPC clusters. This is valid, will pass this check, and will also pass that value down into the deprovisioner.

Reply:
Ack this make sense